### DISCUSSION TITLE
"Remote Url" Based Playlist & Now Playing

### DISCUSSION THREAD

**Discussion by gAlleb** [AzuraCast/AzuraCast#7259] [Ideas]
Could we include `playlist name` into `Now Playing` info when the playlist is "remote url"? Or even make a dedicated field — "remote url playlist" or something. Would be nice. Thank you!

<img width="500" alt="image" src="https://github.com/AzuraCast/AzuraCast/assets/144447266/bec96405-4611-41d2-81d1-db76aed498a7">

**Comment by Moonbase59** [comment]
Indeed, that’d be nice. I miss my nice show names for web site, azuracast_xmltv and video for remote streams… ;-)

![Auswahl_453](https://github.com/AzuraCast/AzuraCast/assets/3706922/3324cd76-5314-417c-a080-14c0e08fb269)

**Comment by BusterNeece** [comment]
I don't believe so, as Liquidsoap doesn't send us info on what playlist triggers a remote URL connection.

**Comment by Moonbase59** [comment]
But doesn’t AzuraCast control the playlists? And thus knows their name?

![Auswahl_454](https://github.com/AzuraCast/AzuraCast/assets/3706922/4c75eb3c-b1fc-4d9a-a9b7-3b403828b45a)

**Reply by BusterNeece** [reply]
You're misunderstanding. In order for us to know what's playing right now, we depend on Liquidsoap feeding that information _back_ to us via its `feedback` mechanism. It's likely that _it_ doesn't know that the remote URL it's playing right now is actually from playlist x in AzuraCast, at least enough to correlate that. In fact, right now it doesn't even send us details about what _URL_ is playing, rather just the title/artist/etc. metadata if that changes.

I'm just not sure if it's at all possible given that limitation.

**Reply by Moonbase59** [reply]
Ah, I see what you’re saying. Since we don’t use a "normal" playlist here (with annotations included) but instead an `input.http`, we only get the remote’s stream data back, like

```
2024/06/30 06:58:52 [metadata_updated:3] ("UTC", "20240630T045843.256")
2024/06/30 06:58:52 [metadata_updated:3] ("icy-br", "128")
2024/06/30 06:58:52 [metadata_updated:3] ("icy-description", "Heart Dance")
2024/06/30 06:58:52 [metadata_updated:3] ("icy-genre", "Pop/Rock")
2024/06/30 06:58:52 [metadata_updated:3] ("icy-name", "Heart Dance")
2024/06/30 06:58:52 [metadata_updated:3] ("icy-private", "0")
2024/06/30 06:58:52 [metadata_updated:3] ("icy-pub", "1")
2024/06/30 06:58:52 [metadata_updated:3] ("title", "Secondcity - I Wanna Feel")
2024/06/30 06:58:52 [metadata_updated:3] ("url", "")
```

Hmm. Let me think…

**Reply by gAlleb** [reply]
@Moonbase59, you do! I wanted to think first, but since you’ve already started, I’ll wait:)))

Maybe we can at least generate an id for input.http and pass the id generated by configwriter.

**Comment by Moonbase59** [comment]
Hm. So I added the `playlist_id` but is doesn’t get picked up. Too bad also we can’t simply [use `annotate` with `input.http`](https://github.com/savonet/liquidsoap/discussions/3998).

```
2024/06/30 08:53:31 [metadata_updated:3] ("UTC", "20240630T065323.397")
2024/06/30 08:53:31 [metadata_updated:3] ("icy-br", "128")
2024/06/30 08:53:31 [metadata_updated:3] ("icy-description", "Heart Dance")
2024/06/30 08:53:31 [metadata_updated:3] ("icy-genre", "Pop/Rock")
2024/06/30 08:53:31 [metadata_updated:3] ("icy-name", "Heart Dance")
2024/06/30 08:53:31 [metadata_updated:3] ("icy-private", "0")
2024/06/30 08:53:31 [metadata_updated:3] ("icy-pub", "1")
2024/06/30 08:53:31 [metadata_updated:3] ("playlist_id", "217")
2024/06/30 08:53:31 [metadata_updated:3] ("title", "David Morales / Juliet Roberts - Needin You")
2024/06/30 08:53:31 [metadata_updated:3] ("url", "")
2024/06/30 08:53:32 [show_meta:3] Now playing: David Morales / Juliet Roberts - Needin You
```

API result for `/nowplaying/niteradio`:
```
  "now_playing": {
    "sh_id": 521870,
    "played_at": 1719730414,
    "duration": 0,
    "playlist": "",
    "streamer": "",
    "is_request": false,
    "song": {
      "id": "762c3567784653c2aa167c17f3f444e3",
      "art": "https://radio.niteradio.net/static/uploads/niteradio/album_art.1706249974.jpg",
      "custom_fields": {
        "bpm": null,
        "year": null,
        "originalyear": null,
        "language": null
      },
      "text": "David Morales / Juliet Roberts - Needin You",
      "artist": "David Morales / Juliet Roberts",
      "title": "Needin You",
      "album": "",
      "genre": "",
      "isrc": "",
      "lyrics": ""
    },
    "elapsed": 72,
    "remaining": 0
  },
```

---

I changed:

Playlist definition, to include the playlist id:
```
def metadata_heart(m) = [...m, ("playlist_id", "217")] end  # ADDED
playlist_heart_dance_from_london2C_uk = mksafe(buffer(buffer=10., input.http(metadata_filter=metadata_heart, "http://example.com/stream")))  # CHANGED
```

`metadata_update`, to include logging metadata, and add `playlist_id` if there is no `song_id`:
```
def metadata_updated(m) =
    def f() =
        label="metadata_updated"  # ADDED
        l = list.sort.natural(metadata.cover.remove(m))  # ADDED
        list.iter(fun(v) -> log(level=3, label=label, "#{v}"), l)  # ADDED

        if (m["is_error_file"] != "true") then
            if (m["title"] != last_title() or m["artist"] != last_artist()) then
                last_title.set(m["title"])
                last_artist.set(m["artist"])

                j = json()

                if (m["song_id"] != "") then
                    j.add("song_id", m["song_id"])
                    j.add("media_id", m["media_id"])
                    j.add("playlist_id", m["playlist_id"])
                else
                    j.add("artist", m["artist"])
                    j.add("title", m["title"])
                    j.add("playlist_id", m["playlist_id"])  # ADDED
                end

                _ = azuracast_api_call(
                    "feedback",
                    json.stringify(j)
                )
            end
        end
    end

    thread.run(f)
end
```
